src/dosext/dpmi/dpmi.c

```diff
@@ -468,7 +468,7 @@ static void dpmi_set_pm(int pm)
   }
   dpmi_pm = pm;
   if (config.cpu_vm != config.cpu_vm_dpmi) {
-    emu_fpstate fpstate;
+    static emu_fpstate fpstate;
```

Comment:
Ahha!
So here's the trick!
Global fpstate IS needed, or you
need hacks like this one.
Please just don't remove the global
fpstate then. Adding private fpstate
to cpu_emu etc is still fine, but please
update the good old `vm86_fpu_state`
here in this func, instead of making
things static.

Reply:
I don't agree about the global fpstate here. It's not like it magically contained more!

The global fpstate also started "static", then got updated at init time with savefpstate from the host (not portable!), which on i386 also did not fill in all the fields! So we had exactly the same issue with global fpstate as with local fpstate!

I don't like the trick either but then the proper solution is that the "get" functions fill in ALL the state and not just parts right?

Reply:
> So we had exactly the same issue with global fpstate as with local fpstate!

No, unless you correct the above
to say "local _static_ fpstate". Then
they are more identical, but still
global state has many advantages,
like being available for save/restore,
being available for reset by 0xf0 and
reboot, and all that.
Neither the global fpstate or static
local fpstate has that crash problem
because they both can accumulate
and keep the proper values from whatever
back-end provides such values. If you
took full fpstate from 1 backend, then
it doesn't matter that you only take part
from back-end 2 - its still fine, the parts
from back-end 1 are preserved.

> I don't like the trick either but then the proper solution is that the "get" functions fill in ALL the state and
> not just parts right?

That's absolutely contrary to what I
did, and I don't see how does this solve
the problem. My idea is trivial (and
hopefully works): just preserve the
right parts of the state if you managed
to take it from either back-end. Get from
another back-end only in what he is
certain to be valid.

Reply:
In fact, even though I am against
having any duality in a global state,
the backend getters can very well
return a flag specifying what state
have they filled. Maybe they can
even be provided with 2 pointers:
one for fxsave and another for
fsave, and they fill up only 1 of
them, saying which one.
Setters should accept any and
locally save the full state even if
they can use only fsave part.
In this case in all back-ends the
full fxsave state will eventually
accumulate, so they will be able
to return it, unless both back-ends
use just fsave.

If we ever want to fill up the global
state in that scheme, then we call
the getter and it will most certainly
return the full fxsave state, as it was
probably "accumulated". If not -
convert to fxsave and store in a
global state. But global state may
not be needed at all.

What do you think?
Or is this what you already done?
Maybe so, as you seem to store
both already...

Reply:
The duality (if needed) is all private at the moment in this PR.
e.g. for true_vm86, "set" stores everything in private fxsave, and converts to fsave if needed.
then it'll use either fsave or fxsave
then with "get", if fsave was used, the fsave fields are merged with the fxsave fields.

Reply:
But the cost of that approach is
to never share the full fxsave?